plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs

```diff
@@ -341,14 +353,46 @@ maybeAddTySubst tn rhs = do
         pure Nothing
     else pure $ Just rhs
 
--- | Is this a an utterly trivial term which might as well be inlined?
-trivialTerm :: Term tyname name uni fun a -> Bool
-trivialTerm = \case
-    Builtin{}  -> True
-    Var{}      -> True
-    -- TODO: Should this depend on the size of the constant?
-    Constant{} -> True
-    _          -> False
+-- | Is the cost increase (in terms of evaluation work) of inlining a variable whose RHS is
+-- the given term acceptable?
+costIsAcceptable :: Term tyname name uni fun a -> Bool
+costIsAcceptable = \case
+  Builtin{}  -> True
+  Var{}      -> True
+  Constant{} -> True
+  Error{}    -> True
+  -- This will mean that we create closures at each use site instead of
+  -- once, but that's a very low cost which we're okay rounding to 0.
+  LamAbs{}   -> True
+  TyAbs{}    -> True
+
+  -- Arguably we could allow these two, but they're uncommon anyway
+  IWrap{}    -> False
+  Unwrap{}   -> False
+  Apply{}    -> False
+  TyInst{}   -> False
+  Let{}      -> False
+
+-- | Is the size increase (in the AST) of inlining a variable whose RHS is
+-- the given term acceptable?
+sizeIsAcceptable :: Term tyname name uni fun a -> Bool
+sizeIsAcceptable = \case
+  Builtin{}      -> True
+  Var{}          -> True
+  Error{}        -> True
+  -- See Note [Inlining criteria]
+  LamAbs _ _ _ t -> sizeIsAcceptable t
+  TyAbs _ _ _ t  -> sizeIsAcceptable t
+
+  -- Arguably we could allow these two, but they're uncommon anyway
+  IWrap{}        -> False
+  Unwrap{}       -> False
+  -- Constants can be big! We could check the size here and inline if they're
+  -- small, but probably not worth it
```

Comment:
@michaelpj This seems well-scoped enough for me to continue my study, will you take an experimental PR? Also, thank you for completing the previous one!

Reply:
I'm pretty sure this isn't worth it. This will only turn up for actual constants bound to variables, which I think is unlikely to be common.

Reply:
I see. I'll flip it to `True` then look at the diff in a few code bases and quit if I can't find anything interesting then.